{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install import_ipynb --no-cache\n",
    "import import_ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[33mWARNING: Ignoring invalid distribution -umpy (/home/rohinsood/.local/lib/python3.10/site-packages)\u001b[0m\u001b[33m\n",
      "\u001b[0m\u001b[33mWARNING: Ignoring invalid distribution -umpy (/home/rohinsood/.local/lib/python3.10/site-packages)\u001b[0m\u001b[33m\n",
      "\u001b[0mRequirement already satisfied: mygene in /home/rohinsood/.local/lib/python3.10/site-packages (3.2.2)\n",
      "Requirement already satisfied: biothings-client>=0.2.6 in /home/rohinsood/.local/lib/python3.10/site-packages (from mygene) (0.3.1)\n",
      "Requirement already satisfied: requests>=2.3.0 in /home/rohinsood/.local/lib/python3.10/site-packages (from biothings-client>=0.2.6->mygene) (2.31.0)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /home/rohinsood/.local/lib/python3.10/site-packages (from requests>=2.3.0->biothings-client>=0.2.6->mygene) (3.2.0)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /home/rohinsood/.local/lib/python3.10/site-packages (from requests>=2.3.0->biothings-client>=0.2.6->mygene) (3.4)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /home/rohinsood/.local/lib/python3.10/site-packages (from requests>=2.3.0->biothings-client>=0.2.6->mygene) (2.0.4)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/rohinsood/.local/lib/python3.10/site-packages (from requests>=2.3.0->biothings-client>=0.2.6->mygene) (2023.7.22)\n",
      "\u001b[33mWARNING: Ignoring invalid distribution -umpy (/home/rohinsood/.local/lib/python3.10/site-packages)\u001b[0m\u001b[33m\n",
      "\u001b[0m\u001b[33mWARNING: Ignoring invalid distribution -umpy (/home/rohinsood/.local/lib/python3.10/site-packages)\u001b[0m\u001b[33m\n",
      "\u001b[0mDefaulting to user installation because normal site-packages is not writeable\n",
      "\u001b[33mWARNING: Ignoring invalid distribution -umpy (/home/rohinsood/.local/lib/python3.10/site-packages)\u001b[0m\u001b[33m\n",
      "\u001b[0m\u001b[33mWARNING: Ignoring invalid distribution -umpy (/home/rohinsood/.local/lib/python3.10/site-packages)\u001b[0m\u001b[33m\n",
      "\u001b[0mRequirement already satisfied: shap in /home/rohinsood/.local/lib/python3.10/site-packages (0.46.0)\n",
      "Requirement already satisfied: numpy in /home/rohinsood/.local/lib/python3.10/site-packages (from shap) (1.26.4)\n",
      "Requirement already satisfied: scipy in /home/rohinsood/.local/lib/python3.10/site-packages (from shap) (1.13.0)\n",
      "Requirement already satisfied: scikit-learn in /home/rohinsood/.local/lib/python3.10/site-packages (from shap) (1.4.2)\n",
      "Requirement already satisfied: pandas in /home/rohinsood/.local/lib/python3.10/site-packages (from shap) (2.2.2)\n",
      "Requirement already satisfied: tqdm>=4.27.0 in /home/rohinsood/.local/lib/python3.10/site-packages (from shap) (4.66.2)\n",
      "Requirement already satisfied: packaging>20.9 in /home/rohinsood/.local/lib/python3.10/site-packages (from shap) (23.1)\n",
      "Requirement already satisfied: slicer==0.0.8 in /home/rohinsood/.local/lib/python3.10/site-packages (from shap) (0.0.8)\n",
      "Requirement already satisfied: numba in /home/rohinsood/.local/lib/python3.10/site-packages (from shap) (0.60.0)\n",
      "Requirement already satisfied: cloudpickle in /home/rohinsood/.local/lib/python3.10/site-packages (from shap) (3.0.0)\n",
      "Requirement already satisfied: llvmlite<0.44,>=0.43.0dev0 in /home/rohinsood/.local/lib/python3.10/site-packages (from numba->shap) (0.43.0)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /home/rohinsood/.local/lib/python3.10/site-packages (from pandas->shap) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in /usr/lib/python3/dist-packages (from pandas->shap) (2022.1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /home/rohinsood/.local/lib/python3.10/site-packages (from pandas->shap) (2024.1)\n",
      "Requirement already satisfied: joblib>=1.2.0 in /home/rohinsood/.local/lib/python3.10/site-packages (from scikit-learn->shap) (1.4.0)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /home/rohinsood/.local/lib/python3.10/site-packages (from scikit-learn->shap) (3.4.0)\n",
      "Requirement already satisfied: six>=1.5 in /usr/lib/python3/dist-packages (from python-dateutil>=2.8.2->pandas->shap) (1.16.0)\n",
      "\u001b[33mWARNING: Ignoring invalid distribution -umpy (/home/rohinsood/.local/lib/python3.10/site-packages)\u001b[0m\u001b[33m\n",
      "\u001b[0m\u001b[33mWARNING: Ignoring invalid distribution -umpy (/home/rohinsood/.local/lib/python3.10/site-packages)\u001b[0m\u001b[33m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "m = __import__(\"Methods\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', None)\n",
    "import matplotlib.pyplot as plt\n",
    "from urllib.request import urlretrieve\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "import numpy as np\n",
    "import ast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Model</th>\n",
       "      <th>Target Variable</th>\n",
       "      <th>MSE</th>\n",
       "      <th>RMSE</th>\n",
       "      <th>MAE</th>\n",
       "      <th>R2</th>\n",
       "      <th>R</th>\n",
       "      <th>Selected Features</th>\n",
       "      <th>Top 20 SHAP Values</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ANN_Final_Test</td>\n",
       "      <td>th_positive_cells</td>\n",
       "      <td>2107.981490</td>\n",
       "      <td>45.912760</td>\n",
       "      <td>25.453749</td>\n",
       "      <td>-0.438400</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FBgn0015777,FBgn0015778,FBgn0015781,FBgn001578...</td>\n",
       "      <td>{'FBgn0023496': 0.03449878091810768, 'FBti0062...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Linear_Regression_Final_Test</td>\n",
       "      <td>th_positive_cells</td>\n",
       "      <td>1977.321753</td>\n",
       "      <td>44.467086</td>\n",
       "      <td>24.632739</td>\n",
       "      <td>-0.349243</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FBgn0000455,FBgn0000459,FBgn0000463,FBgn000055...</td>\n",
       "      <td>{'FBgn0036833': 0.19240042489265521, 'FBgn0264...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Model    Target Variable          MSE       RMSE  \\\n",
       "0                ANN_Final_Test  th_positive_cells  2107.981490  45.912760   \n",
       "1  Linear_Regression_Final_Test  th_positive_cells  1977.321753  44.467086   \n",
       "\n",
       "         MAE        R2   R                                  Selected Features  \\\n",
       "0  25.453749 -0.438400 NaN  FBgn0015777,FBgn0015778,FBgn0015781,FBgn001578...   \n",
       "1  24.632739 -0.349243 NaN  FBgn0000455,FBgn0000459,FBgn0000463,FBgn000055...   \n",
       "\n",
       "                                  Top 20 SHAP Values  \n",
       "0  {'FBgn0023496': 0.03449878091810768, 'FBti0062...  \n",
       "1  {'FBgn0036833': 0.19240042489265521, 'FBgn0264...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_df = pd.read_csv(\"data/results/all_model_results_n1500.csv\")\n",
    "\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "ANN_features = results_df.iloc[0][\"Top 20 SHAP Values\"]\n",
    "ANN_features = ast.literal_eval(ANN_features)\n",
    "df = pd.DataFrame(list(ANN_features.items()), columns=['Gene', 'SHAP Value'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "13 input query terms found no hit:\t['FBti0062398', 'FBti0062381', 'FBti0062695', 'FBti0063086', 'FBti0063079', 'FBti0064274', 'FBti0062\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RNA-seq data shape before:  (20, 2)\n",
      "RNA-seq data shape after:  (7, 2)\n"
     ]
    }
   ],
   "source": [
    "# Import the MyGeneInfo module\n",
    "mg = m.mygene.MyGeneInfo()\n",
    "\n",
    "# Get the list of FlyBase gene IDs from your data\n",
    "flybase_ids = list(df[\"Gene\"])\n",
    "\n",
    "# Query MyGeneInfo for FlyBase IDs\n",
    "ginfo = mg.querymany(flybase_ids, scopes='flybase')\n",
    "\n",
    "# Initialize lists to keep track of genes to drop and those with symbols\n",
    "drop_list = []\n",
    "hit_list = []\n",
    "\n",
    "# Iterate through the query results\n",
    "for g in ginfo:\n",
    "    # Check if the gene has a 'symbol'; if not, add to the drop list\n",
    "    if 'symbol' not in g:\n",
    "        drop_list.append(g['query'])\n",
    "    else:\n",
    "        # Check if the gene symbol is already in the hit list; if not, add it\n",
    "        if g['symbol'] not in hit_list:\n",
    "            # Replace the FlyBase ID with its corresponding gene symbol\n",
    "            df.loc[df['Gene'] == g['query'], 'Gene'] = g['symbol']\n",
    "            hit_list.append(g['symbol'])\n",
    "\n",
    "# Print the shape of the RNA-seq data before filtering\n",
    "print('RNA-seq data shape before: ', df.shape)\n",
    "\n",
    "# Remove rows where the gene ID did not have a corresponding symbol\n",
    "df = df[~df['Gene'].isin(drop_list)]\n",
    "\n",
    "# Print the shape of the RNA-seq data after filtering\n",
    "print('RNA-seq data shape after: ', df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Lip1', 'Ugt37B1', 'Pk34A', 'CG13359', 'Lectin-galC1', 'Or33b', 'ind']\n"
     ]
    }
   ],
   "source": [
    "print(df[\"Gene\"].to_list())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# look at the names of the libraries that are available to use for PEA and GSEA\n",
    "gene_sets = m.gp.get_library_name(organism=\"fly\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Term</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>apterous-expressing neuron of the lateral clus...</td>\n",
       "      <td>ind</td>\n",
       "      <td>0.007866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>apterous-expressing neuron of the lateral clus...</td>\n",
       "      <td>ind</td>\n",
       "      <td>0.007866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ventral_nerve_cord_primordium</td>\n",
       "      <td>ind</td>\n",
       "      <td>0.011188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284</th>\n",
       "      <td>holo TFIIH complex (GO:0005675)</td>\n",
       "      <td>Pk34A</td>\n",
       "      <td>0.025158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293</th>\n",
       "      <td>galactose binding (GO:0005534)</td>\n",
       "      <td>Lectin-galC1</td>\n",
       "      <td>0.018852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294</th>\n",
       "      <td>cyclin-dependent protein kinase activity (GO:0...</td>\n",
       "      <td>Pk34A</td>\n",
       "      <td>0.018852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>cyclin-dependent protein serine/threonine kina...</td>\n",
       "      <td>Pk34A</td>\n",
       "      <td>0.018852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296</th>\n",
       "      <td>lipase activity (GO:0016298)</td>\n",
       "      <td>Lip1</td>\n",
       "      <td>0.042963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>olfactory receptor activity (GO:0004984)</td>\n",
       "      <td>Or33b</td>\n",
       "      <td>0.044263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>protein_tyrosine_kinase_binding_(GO:1990782)</td>\n",
       "      <td>ind</td>\n",
       "      <td>0.028695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311</th>\n",
       "      <td>MAP_kinase_kinase_kinase_activity_(GO:0004709)</td>\n",
       "      <td>ind</td>\n",
       "      <td>0.028695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312</th>\n",
       "      <td>nucleotide_kinase_activity_(GO:0019201)</td>\n",
       "      <td>ind</td>\n",
       "      <td>0.028695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>Nup54</td>\n",
       "      <td>Pk34A</td>\n",
       "      <td>0.023988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>316</th>\n",
       "      <td>Neos</td>\n",
       "      <td>Pk34A</td>\n",
       "      <td>0.023988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>317</th>\n",
       "      <td>CG18745</td>\n",
       "      <td>Pk34A</td>\n",
       "      <td>0.047654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>axis_specification_defective</td>\n",
       "      <td>ind</td>\n",
       "      <td>0.022350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>322</th>\n",
       "      <td>axis_specification_defective</td>\n",
       "      <td>ind</td>\n",
       "      <td>0.002797</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  Term         Genes  \\\n",
       "0    apterous-expressing neuron of the lateral clus...           ind   \n",
       "1    apterous-expressing neuron of the lateral clus...           ind   \n",
       "24                       ventral_nerve_cord_primordium           ind   \n",
       "284                    holo TFIIH complex (GO:0005675)         Pk34A   \n",
       "293                     galactose binding (GO:0005534)  Lectin-galC1   \n",
       "294  cyclin-dependent protein kinase activity (GO:0...         Pk34A   \n",
       "295  cyclin-dependent protein serine/threonine kina...         Pk34A   \n",
       "296                       lipase activity (GO:0016298)          Lip1   \n",
       "297           olfactory receptor activity (GO:0004984)         Or33b   \n",
       "310       protein_tyrosine_kinase_binding_(GO:1990782)           ind   \n",
       "311     MAP_kinase_kinase_kinase_activity_(GO:0004709)           ind   \n",
       "312            nucleotide_kinase_activity_(GO:0019201)           ind   \n",
       "315                                              Nup54         Pk34A   \n",
       "316                                               Neos         Pk34A   \n",
       "317                                            CG18745         Pk34A   \n",
       "318                       axis_specification_defective           ind   \n",
       "322                       axis_specification_defective           ind   \n",
       "\n",
       "     Adjusted P-value  \n",
       "0            0.007866  \n",
       "1            0.007866  \n",
       "24           0.011188  \n",
       "284          0.025158  \n",
       "293          0.018852  \n",
       "294          0.018852  \n",
       "295          0.018852  \n",
       "296          0.042963  \n",
       "297          0.044263  \n",
       "310          0.028695  \n",
       "311          0.028695  \n",
       "312          0.028695  \n",
       "315          0.023988  \n",
       "316          0.023988  \n",
       "317          0.047654  \n",
       "318          0.022350  \n",
       "322          0.002797  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Perform enrichment analysis with FlyBase genes using Enrichr\n",
    "enr = m.gp.enrichr(\n",
    "    gene_list=df[\"Gene\"].to_list(),  # List of FlyBase gene symbols\n",
    "    gene_sets=gene_sets,\n",
    "    organism='fly',  # Set to 'fly' for Drosophila (FlyBase genes)\n",
    "    outdir=None\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genes</th>\n",
       "      <th>Term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ind</td>\n",
       "      <td>apterous-expressing neuron of the lateral cluster of the thorax Ap4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ind</td>\n",
       "      <td>apterous-expressing neuron of the lateral cluster of the thorax</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ind</td>\n",
       "      <td>ventral_nerve_cord_primordium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284</th>\n",
       "      <td>Pk34A</td>\n",
       "      <td>holo TFIIH complex (GO:0005675)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293</th>\n",
       "      <td>Lectin-galC1</td>\n",
       "      <td>galactose binding (GO:0005534)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294</th>\n",
       "      <td>Pk34A</td>\n",
       "      <td>cyclin-dependent protein kinase activity (GO:0097472)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>Pk34A</td>\n",
       "      <td>cyclin-dependent protein serine/threonine kinase activity (GO:0004693)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296</th>\n",
       "      <td>Lip1</td>\n",
       "      <td>lipase activity (GO:0016298)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>Or33b</td>\n",
       "      <td>olfactory receptor activity (GO:0004984)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>ind</td>\n",
       "      <td>protein_tyrosine_kinase_binding_(GO:1990782)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311</th>\n",
       "      <td>ind</td>\n",
       "      <td>MAP_kinase_kinase_kinase_activity_(GO:0004709)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312</th>\n",
       "      <td>ind</td>\n",
       "      <td>nucleotide_kinase_activity_(GO:0019201)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>Pk34A</td>\n",
       "      <td>Nup54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>316</th>\n",
       "      <td>Pk34A</td>\n",
       "      <td>Neos</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>317</th>\n",
       "      <td>Pk34A</td>\n",
       "      <td>CG18745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>ind</td>\n",
       "      <td>axis_specification_defective</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>322</th>\n",
       "      <td>ind</td>\n",
       "      <td>axis_specification_defective</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Genes  \\\n",
       "0             ind   \n",
       "1             ind   \n",
       "24            ind   \n",
       "284         Pk34A   \n",
       "293  Lectin-galC1   \n",
       "294         Pk34A   \n",
       "295         Pk34A   \n",
       "296          Lip1   \n",
       "297         Or33b   \n",
       "310           ind   \n",
       "311           ind   \n",
       "312           ind   \n",
       "315         Pk34A   \n",
       "316         Pk34A   \n",
       "317         Pk34A   \n",
       "318           ind   \n",
       "322           ind   \n",
       "\n",
       "                                                                       Term  \n",
       "0       apterous-expressing neuron of the lateral cluster of the thorax Ap4  \n",
       "1           apterous-expressing neuron of the lateral cluster of the thorax  \n",
       "24                                            ventral_nerve_cord_primordium  \n",
       "284                                         holo TFIIH complex (GO:0005675)  \n",
       "293                                          galactose binding (GO:0005534)  \n",
       "294                   cyclin-dependent protein kinase activity (GO:0097472)  \n",
       "295  cyclin-dependent protein serine/threonine kinase activity (GO:0004693)  \n",
       "296                                            lipase activity (GO:0016298)  \n",
       "297                                olfactory receptor activity (GO:0004984)  \n",
       "310                            protein_tyrosine_kinase_binding_(GO:1990782)  \n",
       "311                          MAP_kinase_kinase_kinase_activity_(GO:0004709)  \n",
       "312                                 nucleotide_kinase_activity_(GO:0019201)  \n",
       "315                                                                   Nup54  \n",
       "316                                                                    Neos  \n",
       "317                                                                 CG18745  \n",
       "318                                            axis_specification_defective  \n",
       "322                                            axis_specification_defective  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.set_option('display.max_colwidth', None)\n",
    "enr.results[enr.results['Adjusted P-value'] < 0.05][[\"Genes\", 'Term']]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0]"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "767d51c1340bd893661ea55ea3124f6de3c7a262a8b4abca0554b478b1e2ff90"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
